File: apps/apple/core/Sources/LorvexStore/FocusScheduleSnapshot.swift
import Foundation
import GRDB
import LorvexDomain

/// Shared focus-schedule block normalization for sync / export payloads.
///
/// Provenance is carried by each block, so serialization is independent of
/// calendar-row arrival order. Provider titles are device-local calendar data
/// and are replaced with `"Event"` before sync or export; canonical and
/// freeform block titles are preserved.
public enum FocusScheduleSnapshot {
  /// Neutral label substituted for a provider-derived event block's title.
  static let neutralEventTitle = "Event"

  /// Apply the external-transfer privacy rule to a single block. Provider
  /// blocks never carry identity and always use the neutral title.
  public static func normalizeBlockForExternalTransfer(
    eventSource: FocusScheduleEventSource?, calendarEventId: String?, title: String?
  ) -> (calendarEventId: String?, title: String?) {
    guard eventSource == .provider else { return (calendarEventId, title) }
    return (nil, neutralEventTitle)
  }

  /// Serialize the `focus_schedule_blocks` rows for `date` into the
  /// canonical JSON array embedded in a `focus_schedule` sync envelope.
  /// Position-ordered; each block carries `block_type`, `start_minutes`,
  /// `end_minutes` (integer minute offsets), `task_id`, `calendar_event_id`,
  /// `event_source`, `title`, after the provider-block normalization above.
  public static func serializeBlocksForSync(
    _ db: Database, date: String
  ) throws -> [JSONValue] {
    let rows = try Row.fetchAll(
      db,
      sql: """
        SELECT block_type, start_minutes, end_minutes, task_id, calendar_event_id, event_source, title \
        FROM focus_schedule_blocks \
        WHERE date = ? ORDER BY position ASC
        """,
      arguments: [date])

    var blocks: [JSONValue] = []
    blocks.reserveCapacity(rows.count)
    for row in rows {
      let blockType: String = row["block_type"]
      let startMinutes: Int64 = row["start_minutes"]
      let endMinutes: Int64 = row["end_minutes"]
      let taskId: String? = row["task_id"]
      let calendarEventId: String? = row["calendar_event_id"]
      let eventSourceRaw: String? = row["event_source"]
      let title: String? = row["title"]
      let eventSource = eventSourceRaw.flatMap(FocusScheduleEventSource.parse)

      let (syncedCalendarEventId, syncedTitle) = normalizeBlockForExternalTransfer(
        eventSource: eventSource, calendarEventId: calendarEventId, title: title)

      blocks.append(
        .object([
          "block_type": .string(blockType),
          "start_minutes": .int(startMinutes),
          "end_minutes": .int(endMinutes),
          "task_id": taskId.map(JSONValue.string) ?? .null,
          "calendar_event_id": syncedCalendarEventId.map(JSONValue.string) ?? .null,
          "event_source": eventSourceRaw.map(JSONValue.string) ?? .null,
          "title": syncedTitle.map(JSONValue.string) ?? .null,
        ]))
    }
    return blocks
  }
}

File: apps/tauri/lorvex-store/src/focus_schedule_snapshot.rs
//! Shared focus schedule block normalization for sync/export.
//!
//! Provider-derived event blocks must have their event_id stripped and
//! title neutralized to "Event" before entering any synced payload
//! (doc 19 write-down rule).

use rusqlite::Connection;
use serde_json::Value;

/// Neutral label for provider-derived event blocks in synced payloads.
const NEUTRAL_EVENT_TITLE: &str = "Event";

/// Normalize a single focus schedule block for sync.
///
/// If `is_canonical_event` is false and the block has an event_id,
/// the event_id is stripped and the title neutralized to "Event".
///
/// This is the shared normalization contract. All serialization paths
/// (DB-row based: export/reseed, in-memory struct based: live enqueue/MCP)
/// must call this or apply equivalent logic.
fn normalize_block_for_sync(
    event_id: Option<String>,
    title: Option<String>,
    is_canonical_event: bool,
) -> (Option<String>, Option<String>) {
    let is_provider_block = event_id.is_some() && !is_canonical_event;
    let synced_event_id = if is_provider_block { None } else { event_id };
    let synced_title = if is_provider_block {
        Some(NEUTRAL_EVENT_TITLE.to_string())
    } else {
        title
    };
    (synced_event_id, synced_title)
}

/// Serialize focus_schedule blocks for a given date into a canonical
/// JSON array suitable for sync/export payloads.
///
/// Uses `normalize_block_for_sync` for each block, with a LEFT JOIN
/// to determine canonical status.
pub fn serialize_blocks_for_sync(
    conn: &Connection,
    schedule_date: &str,
) -> Result<Vec<Value>, rusqlite::Error> {
    let mut stmt = conn.prepare_cached(
        "SELECT b.block_type, b.start_time, b.end_time, b.task_id, b.event_id, b.title,
                ce.id AS canonical_event_id
         FROM focus_schedule_blocks b
         LEFT JOIN calendar_events ce ON ce.id = b.event_id
         WHERE b.schedule_date = ?1 ORDER BY b.position ASC",
    )?;

    let mut blocks = Vec::new();
    let mut rows = stmt.query(rusqlite::params![schedule_date])?;
    while let Some(row) = rows.next()? {
        let block_type: String = row.get(0)?;
        let start_time: i64 = row.get(1)?;
        let end_time: i64 = row.get(2)?;
        let task_id: Option<String> = row.get(3)?;
        let event_id: Option<String> = row.get(4)?;
        let title: Option<String> = row.get(5)?;
        let is_canonical: Option<String> = row.get(6)?;

        let (synced_event_id, synced_title) =
            normalize_block_for_sync(event_id, title, is_canonical.is_some());

        blocks.push(serde_json::json!({
            "block_type": block_type,
            "start_time": start_time,
            "end_time": end_time,
            "task_id": task_id,
            "event_id": synced_event_id,
            "title": synced_title,
        }));
    }

    Ok(blocks)
}
